src/liballoc/vec.rs

```diff
@@ -290,6 +290,7 @@ use crate::raw_vec::RawVec;
 /// [`reserve`]: ../../std/vec/struct.Vec.html#method.reserve
 /// [owned slice]: ../../std/boxed/struct.Box.html
 #[stable(feature = "rust1", since = "1.0.0")]
+#[clippy::diagnostic = "vec_type"]
```

Comment:
note that since it's perfectly legal for users to use `#[clippy::diagnostic = "vec_type"]` and only clippy will complain to them, they will get an error with the current impl. I believe that we can do this in a way, that only clippy reports diagnostics about users using `clippy::diagnostic` with the same name as something declared in libstd/libcore.

The way we'd do this is to add an unstable `#[diagnostic_item = "vec_type"]` scheme, and have clippy overwrite the `get_diagnostic_items` query to also inject everything marked with `clippy::diagnostic`. So without the clippy driver there'd only be the unstable scheme, but with clippy, we'd error out because rustc would not be able to differentiate between `diagnostic_item` and `clippy::diagnostic`.

Before I implement this I just want to make sure this is how we wanna roll this.

The reason I want users to be able to use `clippy::diagnostic` is that we have lints for e.g. `serde` and we'd want to annotate `trait Serialize` with `#[clippy::diagnostic = "serde_serialize"]` cc @dtolnay